### PR TITLE
Added cancelation to tile getter interface.

### PR DESCRIPTION
**Pull request by vmihaylenko**
It's sometimes helpful for users to have an option to interrupt network requests. This pr introduces the simple solution for it: passing a function which will be called (if set) to check if `tile_getter_t` should interrupt the particular url fetching. For the default implementation (curl based) we have a callback which is called (if set) each time the progress function is called.
So a possible solution for users who want to use default curl-based tile getter and who also want to interrupt request would be having an `atomic<bool>` which returns from a custom interrupt function the user provided, or, if the user wants to interrupt the particular url, adding the url in a thread safe collection of urls and check that the collection contains url inside the function.

### CODE REVIEW

File: valhalla/baldr/curl_tilegetter.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+#include <string>
+#include <utility>
+
+namespace valhalla {
+namespace baldr {
+
+/**
+ * Default implementation which uses libcurl and curler_pool_t.
+ */
+class curl_tile_getter_t : public tile_getter_t {
```
Comment (mandeepsandhu) [inline valhalla/baldr/curl_tilegetter.h:15]: The convention for class names in valhalla seem to be to use camel-case. Any particular reason this (and the class it derives from) uses snake-case instead? It would be good to standardize around a single convention for better readability.

Comment (kevinkreiser) [reply valhalla/baldr/curl_tilegetter.h:15]: @mandeepsandhu thats not really true. its true in some namespaces (especially Odin) but less true in other ones.

Comment (vmihaylenko) [reply valhalla/baldr/curl_tilegetter.h:15]: Frankly by the moment I wrote the base and derived I'd already seen many classes named in snake case here. Thanks for noticing, going to rewrite everything in camel case.

File: test/http_tiles.cc [hunk 2]
```diff
@@ -258,6 +261,64 @@ TEST(HttpTiles, test_graphreader_multiple_threads) {
...
+        if (!is_for_this_thread) {
+          continue;
+        }
+
+        auto test_tile_index = tile_i % params.test_tile_names.size();
+        auto expected_tile_id = params.test_tile_ids[test_tile_index];
+
+        auto tile_uri = url_builder(tile_i);
+        if (canceled_uris.find(tile_uri) != canceled_uris.end()) {
+          EXPECT_THROW(tile_getter.get(tile_uri), std::runtime_error);
```
Comment (mskurydin) [inline test/http_tiles.cc:302]: in this case it seems to throw before event starting network activity?

File: valhalla/baldr/graphreader.h [hunk 3]
```diff
@@ -343,6 +343,12 @@ class GraphReader {
   explicit GraphReader(const boost::property_tree::ptree& pt,
                        std::unique_ptr<tile_getter_t>&& tile_getter = nullptr);
 
+  void set_tile_getter_interrupt(const tile_getter_t::interrupt_t* interrupt) {
+    if (tile_getter_) {
+      tile_getter_->set_interrupt(interrupt);
```
Comment (mskurydin) [inline valhalla/baldr/graphreader.h:348]: what is the process of cancelling everything from a higher level?
suppose we are building a route with thor which issues a blocking tile download.
if we want to abort all the activity, do we wrap always 'true' function and set it with set_tile_getter_interrupt?

File: valhalla/baldr/graphreader.h [hunk 4]
```diff
@@ -343,6 +343,12 @@ class GraphReader {
   explicit GraphReader(const boost::property_tree::ptree& pt,
                        std::unique_ptr<tile_getter_t>&& tile_getter = nullptr);
 
+  void set_tile_getter_interrupt(const tile_getter_t::interrupt_t* interrupt) {
```
Comment (mskurydin) [inline valhalla/baldr/graphreader.h:346]: The naming style is a bit off compared to graph reader class (which is more of pure google c++ code style)

File: valhalla/baldr/tilegetter.h [hunk 5]
```diff
@@ -45,46 +45,18 @@ class tile_getter_t {
...
-   * @param user_agent  user agent to use by curlers for HTTP requests
-   * @param gzipped  whether to request for gzip compressed data
+   * A callback which is called to check if the request should be interrupted.
    */
-  curl_tile_getter_t(const size_t pool_size, const std::string& user_agent, bool gzipped)
-      : curlers_(pool_size, user_agent), gzipped_(gzipped) {
-  }
-
-  using response_t = tile_getter_t::response_t;
+  using interrupt_t = std::function<bool(const std::string&)>;
```
Comment (mskurydin) [inline valhalla/baldr/tilegetter.h:51]: the network requests are initiated pretty deep in the call stack. If we want to abort a routing, we don't know which URLs are being fetched at the moment. Not sure if we can find a valid use case for selective cancel.

**Comment by mskurydin** [comment]
If we use one GraphReader per function (instead of sharing it with all) and rely on shared cache, we may easily kill all requests since no concurrent downloads for different callers will be taking place.

File: valhalla/baldr/graphreader.h (hunk 3 above)
Comment (vmihaylenko) [reply valhalla/baldr/graphreader.h:348]: As I see it it will be something like 
```cpp
class WorkerWrapper {
public:
   WorkerWrapper() {
      interrupt_ = [this] (string) { return shouldInterrupt_; };
   }

  ~WorkerWrapper() {
      shouldInterrupt_ = {true};
   }

   void setupWorkers() {
      GraphReader reader;
      reader->set_interrupt(&interrupt_);
      boost::property_tree::ptree& conf;
      worker_ = make_unique<service_worker_t>(conf, reader);
   }
   
   void doWork() {
      try {
         worker_->route("");
      } catch (...) {
         // if we cancelled the fetch request we should fall here
      }
   }
  
private:
   atomic<bool> shouldInterrupt_{false};
   function<bool(string)> interrupt;
   unique_ptr<service_worker_t> worker_;
};
```

File: test/http_tiles.cc (hunk 2 above)
Comment (vmihaylenko) [reply test/http_tiles.cc:302]: not really, we will have first throw only after first progress is received.

File: valhalla/baldr/graphreader.h (hunk 4 above)
Comment (vmihaylenko) [reply valhalla/baldr/graphreader.h:346]: oh I see, thanks for pointing me out
though the code style of namespace baldr is a little bit confusing :/

File: valhalla/baldr/tilegetter.h (hunk 5 above)
Comment (vmihaylenko) [reply valhalla/baldr/tilegetter.h:51]: I thought about it but ended up with the idea that having an opportunity to cancel the particular url could be useful and yet could be not usable if we are willing to cancel everything (just don't check the url in this case and return true for everything). The tile getter works with urls and quiet abstract from building routes, searching tiles etc.
But I'm open to this suggestion if you think that would be better. Even more, it would be easier to adapt a function without parameters for cascading cancelation.

File: valhalla/baldr/curl_tilegetter.h (hunk 1 above)
Comment (mandeepsandhu) [reply valhalla/baldr/curl_tilegetter.h:15]: Thats unfortunate. Yeah, I just spot checked only a few modules in `baldr` & `odin` to see if all classes were camel-cased.

So is there a preferred convention? (it seems to be to follow whatever existing convention is used in a file?)

**Comment by vmihaylenko** [comment]
@kevinkreiser @mandeepsandhu ptal

File: valhalla/baldr/graphreader.h (hunk 3 above)
Comment (kevinkreiser) [reply valhalla/baldr/graphreader.h:348]: Yeah this was my main question all along. At the moment the top level interrupts for the various workers can only be set via the http interface. we'd need to allow them being set separately and we'd need to also some how tile all the outstanding tile fetches to a particular request (in the case of multi threaded workers). i had talked to @vmihaylenko a few days ago about exactly this problem and i think its ok to solve this in a separate PR.

in summary and more specifically, id like to honor the existing interrupt structures we have for the top level worker actions and upon them triggering, id like to cascade that interrupt down to any in progress tile fetching....

actually i might have an idea here that might be a bit of a re-write of this work. let me see if i can formulate it into something actionable.

File: valhalla/baldr/graphreader.h [hunk 6]
```diff
@@ -343,6 +343,12 @@ class GraphReader {
   explicit GraphReader(const boost::property_tree::ptree& pt,
                        std::unique_ptr<tile_getter_t>&& tile_getter = nullptr);
 
+  void SetTileGetterInterrupt(const tile_getter_t::interrupt_t* interrupt) {
```
Comment (kevinkreiser) [inline valhalla/baldr/graphreader.h:346]: can we name this to SetInterrupt so it looks like the rest of the workers. its ok that only the tile_getter supports being interrupted, but in the future we might want to interrupt any other underlying storage fetcher

Comment (vmihaylenko) [reply valhalla/baldr/graphreader.h:346]: sure
